Strange behavior with PyPy in 0.10 prerelease

Please provide the following information: 
 * Your python version (`python -V`): PyPy 7.3.3 with Python (3.6.12)
 * Your pip version (`pip -V`): 
   * If this isn't the [latest pip version](https://pip.pypa.io/en/stable/news/), have you checked using the latest pip version?
 * The version of the bindings you're using, if any (e.g. `pyo3`, `rust-cpython` or `cffi`): pyo3
 * Does `cargo build` work? I think so
 * If on windows, have you checked that you aren't accidentally using unix path (those with the forward slash `/`)?

Please list the exact steps required to reproduce your error with all command output and if possible with a repository:

maturin-0.10.0b6 , https://github.com/milesgranger/pyrus-cramjam/runs/2218894957?check_suite_focus=true errored with

```bash
maturin build -i /opt/hostedtoolcache/PyPy/3.6.12/x64/bin/pypy --release --out dist
🔗 Found pyo3 bindings
🐍 Found PyPy 3.6 at /opt/hostedtoolcache/PyPy/3.6.12/x64/bin/pypy
📦 Built source distribution to dist/cramjam-2.1.0.tar.gz
warning: version requirement `0.6.0+zstd.1.4.8` for dependency `zstd` includes semver metadata which will be ignored, removing the metadata is recommended to avoid confusion
   Compiling cfg-if v1.0.0
   Compiling scopeguard v1.1.0
error[E0463]: can't find crate for `core`
  |
  = note: the `x86_64-unknown-linux-musl` target may not be installed


error: aborting due to previous error


error[E0463]: can't find crate for `core`
  |
  = note: the `x86_64-unknown-linux-musl` target may not be installed


error: aborting due to previous error


error: could not compile `scopeguard`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
💥 maturin failed
  Caused by: Failed to build a native library through cargo
  Caused by: Cargo build finished with "exit code: 101": `cargo rustc --message-format json --manifest-path Cargo.toml --release --target x86_64-unknown-linux-musl --lib --`
```

It's strange that `--target x86_64-unknown-linux-musl` is passed to `cargo rustc`, it works fine with maturin 0.9.4

Also causing issues in CPython builds for `rust-numpy`: https://github.com/PyO3/rust-numpy/pull/179/checks?check_run_id=2260903749

```
poetry run maturin develop
🔗 Found pyo3 bindings
🐍 Found CPython 3.8 at python
   Compiling cfg-if v1.0.0
   Compiling smallvec v1.6.1
error[E0463]: can't find crate for `core`
  |
  = note: the `x86_64-unknown-linux-musl` target may not be installed


error[E0463]: can't find crate for `core`
  |
  = note: the `x86_64-unknown-linux-musl` target may not be installed


error: aborting due to previous error


error: aborting due to previous error


error: could not compile `cfg-if`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
💥 maturin failed
  Caused by: Failed to build a native library through cargo
  Caused by: Failed to build a native library through cargo
  Caused by: Cargo build finished with "exit code: 101": `cargo rustc --message-format json --manifest-path Cargo.toml --target x86_64-unknown-linux-musl --lib --`
Error: Process completed with exit code 1.
```

Strangely, if I run the above with a locally built (debug) version of maturin (tag `v0.10.0`) it succeeds. I've instrumented this to print out version info and musl is not mentioned:

```diff
diff --git a/src/target.rs b/src/target.rs
index f875c44..4b6623e 100644
--- a/src/target.rs
+++ b/src/target.rs
@@ -107,6 +107,9 @@ impl Target {
             (Os::Windows, Arch::Armv7L) => bail!("armv7l is not supported for Windows"),
             (_, _) => {}
         }
+
+        println!("Created target: {:?} {:?}, {:?}: {:?}", os, arch, platform.target_env, platform.target_triple);
+
         Ok(Target {
             os,
             arch,
```

```
Created target: Linux X86_64, Some(GNU): "x86_64-unknown-linux-gnu"
```

Also works with `v0.9.4` from pip.

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -76,7 +76,7 @@ jobs:
 
       - name: Install musl tools
         if: matrix.os == 'ubuntu-latest'
-        run: sudo apt install -y musl-tools
+        run: sudo apt-get install -y musl-tools
 
       # Those two will also create target/${{ matrix.target }}/maturin
       - name: Publish to pypi (with sdist)

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -43,6 +43,9 @@ jobs:
       - name: Install x86_64-unknown-linux-musl Rust target
         if: matrix.os == 'ubuntu-latest'
         run: rustup target add x86_64-unknown-linux-musl
+      - name: Install musl tools
+        if: matrix.os == 'ubuntu-latest'
+        run: sudo apt-get install -y musl-tools
       #- name: Cache cargo registry
       #  uses: actions/cache@v2
       #  with:
@@ -72,6 +75,12 @@ jobs:
         uses: actions-rs/cargo@v1
         with:
           command: test
+      - name: cargo test with musl
+        if: matrix.os == 'ubuntu-latest'
+        uses: actions-rs/cargo@v1
+        with:
+          command: test
+          args: --target x86_64-unknown-linux-musl
 
   test-auditwheel:
     name: Test Auditwheel

--- a/.gitignore
+++ b/.gitignore
@@ -10,3 +10,4 @@ __pycache__/
 dist/
 build
 dist
+tags

--- a/src/compile.rs
+++ b/src/compile.rs
@@ -22,12 +22,7 @@ pub fn compile(
     if context.target.is_macos() && context.universal2 {
         compile_universal2(context, python_interpreter, bindings_crate)
     } else {
-        compile_target(
-            context,
-            python_interpreter,
-            bindings_crate,
-            context.target.target_triple(),
-        )
+        compile_target(context, python_interpreter, bindings_crate, None)
     }
 }
 
@@ -45,7 +40,7 @@ fn compile_universal2(
         context,
         python_interpreter,
         bindings_crate,
-        "aarch64-apple-darwin",
+        Some("aarch64-apple-darwin"),
     )
     .context("Failed to build a aarch64 library through cargo")?
     .get(build_type)
@@ -64,7 +59,7 @@ fn compile_universal2(
         context,
         python_interpreter,
         bindings_crate,
-        "x86_64-apple-darwin",
+        Some("x86_64-apple-darwin"),
     )
     .context("Failed to build a x86_64 library through cargo")?
     .get(build_type)
@@ -107,7 +102,7 @@ fn compile_target(
     context: &BuildContext,
     python_interpreter: Option<&PythonInterpreter>,
     bindings_crate: &BridgeModel,
-    target: &str,
+    target: Option<&str>,
 ) -> Result<HashMap<String, PathBuf>> {
     let mut shared_args = vec!["--manifest-path", context.manifest_path.to_str().unwrap()];
 
@@ -116,8 +111,10 @@ fn compile_target(
     if context.release {
         shared_args.push("--release");
     }
-    shared_args.push("--target");
-    shared_args.push(target);
+    if let Some(target) = target {
+        shared_args.push("--target");
+        shared_args.push(target);
+    }
 
     let mut rustc_args: Vec<&str> = context
         .rustc_extra_args

--- a/src/target.rs
+++ b/src/target.rs
@@ -1,5 +1,5 @@
 use crate::{Manylinux, PythonInterpreter};
-use anyhow::{bail, format_err, Result};
+use anyhow::{bail, format_err, Context, Result};
 use platform_info::*;
 use platforms::target::Env;
 use platforms::Platform;
@@ -7,6 +7,8 @@ use std::env;
 use std::fmt;
 use std::path::Path;
 use std::path::PathBuf;
+use std::process::Command;
+use std::str;
 
 /// All supported operating system
 #[derive(Debug, Clone, Eq, PartialEq)]
@@ -47,7 +49,6 @@ pub struct Target {
     os: Os,
     arch: Arch,
     env: Option<Env>,
-    triple: &'static str,
 }
 
 impl Target {
@@ -60,8 +61,9 @@ impl Target {
             Platform::find(target_triple)
                 .ok_or_else(|| format_err!("Unknown target triple {}", target_triple))?
         } else {
-            Platform::guess_current()
-                .ok_or_else(|| format_err!("Could guess the current platform"))?
+            let target_triple = get_host_target()?;
+            Platform::find(&target_triple)
+                .ok_or_else(|| format_err!("Unknown target triple {}", target_triple))?
         };
 
         let os = match platform.target_os {
@@ -111,7 +113,6 @@ impl Target {
             os,
             arch,
             env: platform.target_env,
-            triple: platform.target_triple,
         })
     }
 
@@ -130,11 +131,6 @@ impl Target {
     /// Returns target architecture
     pub fn target_arch(&self) -> Arch {
         self.arch
-    }
-
-    /// Returns target triple string
-    pub fn target_triple(&self) -> &str {
-        &self.triple
     }
 
     /// Returns true if the current platform is linux or mac os
@@ -315,4 +311,27 @@ impl Target {
         let tags = self.get_py3_tags(&manylinux, universal2);
         (tag, tags)
     }
+}
+
+fn get_host_target() -> Result<String> {
+    let output = Command::new("rustc")
+        .arg("-vV")
+        .output()
+        .context("Failed to run rustc to get the host target")?;
+    let output = str::from_utf8(&output.stdout).context("`rustc -vV` didn't return utf8 output")?;
+
+    let field = "host: ";
+    let host = output
+        .lines()
+        .find(|l| l.starts_with(field))
+        .map(|l| &l[field.len()..])
+        .ok_or_else(|| {
+            format_err!(
+                "`rustc -vV` didn't have a line for `{}`, got:\n{}",
+                field.trim(),
+                output
+            )
+        })?
+        .to_string();
+    Ok(host)
 }

